src/System.Runtime.Numerics/src/System/Numerics/BigCalc.SquMul.cs

```diff
@@ -0,0 +1,356 @@
+﻿using System.Diagnostics;
+using System.Security;
+
+namespace System.Numerics
+{
+    internal static partial class BigCalc
+    {
+#if DEBUG
+
+        private const int SquareThreshold = 128 / 32;
+        private const int MultiplyThreshold = 128 / 32;
+
+#else
+
+        private const int SquareThreshold = 1024 / 32;
+        private const int MultiplyThreshold = 1024 / 32;
+
+#endif
+
+        public static uint[] Square(uint[] value)
+        {
+            Debug.Assert(value != null);
+
+            return Square(value, value.Length);
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static uint[] Square(uint[] value, int valueLength)
+        {
+            Debug.Assert(value != null);
+            Debug.Assert(value.Length >= valueLength);
+
+            var bits = new uint[valueLength * 2];
+
+            fixed (uint* v = value, b = bits)
+            {
+                Square(v, valueLength,
+                       b, bits.Length);
+            }
+
+            return bits;
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static void Square(uint* value, int valueLength,
+                                          uint* bits, int bitsLength)
+        {
+            Debug.Assert(valueLength >= 0);
+            Debug.Assert(bitsLength == valueLength * 2);
+
+            if (valueLength < SquareThreshold)
+            {
+                // squares the bits
+                for (var i = 0; i < valueLength; i++)
+                {
+                    var carry = 0UL;
+                    for (var j = 0; j < i; j++)
+                    {
+                        var digit1 = bits[i + j] + carry;
+                        var digit2 = (ulong)value[j] * value[i];
+                        bits[i + j] = (uint)(digit1 + (digit2 << 1));
+                        carry = (digit2 + (digit1 >> 1)) >> 31;
+                    }
+                    var digits = (ulong)value[i] * value[i] + carry;
+                    bits[i * 2] = (uint)digits;
+                    bits[i * 2 + 1] = (uint)(digits >> 32);
+                }
+            }
+            else
+            {
+                // divide & conquer
+                // (see multiply...)
+                var n = valueLength / 2;
```

Comment:
Can be a shift too. I thought that every compiler optimizes `*2` / `/2` using shifts. Thus, writing it that way is preferred, because of readability. (?)